File: LeetCode/0633-sum-of-square-numbers/0633-sum-of-square-numbers.cpp
class Solution {
public:
    bool judgeSquareSum(int c) {
        long a = 0, b = sqrt(c);
        while (a <= b) {
            long k = a * a + b * b;
            if (k < c) {
                a++;
            } else if (k > c) {
                b--;
            } else {
                return true;
            }
        }
        return false;
    }
};



// class Solution {
// public:
//     bool judgeSquareSum(int c) {
//         for (long i = 0; i * i <= c; i++) {
//             int x = c - (int)(i * i);
//             int start = 0, end = x;
//             while (start <= end) {
//                 long mid = start + (end - start) / 2;
//                 if (mid * mid == x) {
//                     return true;
//                 } else if (mid * mid > x) {
//                     end = mid - 1;
//                 } else {
//                     start = mid + 1;
//                 }
//             }
//         }
//         return false;
//     }
// };



// class Solution {
// public:
//     bool judgeSquareSum(int c) {
//         for (long i = 0; i * i <= c; i++) {
//             double b = sqrt(c - i * i);
//             if (b == (int)b) {
//                 return true;
//             }
//         }
//         return false;
//     }
// };



// class Solution {
// public:
//     bool judgeSquareSum(int c) {
//         for (int i = 0; i * i <= c; i++) {
//             for (int j = 0; j * j <= c; j++) {
//                 if (i * i + j * j == c) {
//                     return true;
//                 }
//             }
//         }
//         return false;
//     }
// };

File: LeetCode/0633-sum-of-square-numbers/0633-sum-of-square-numbers.py
class Solution:
    def judgeSquareSum(self, c: int) -> bool:
        a = 0
        b = int(math.sqrt(c))

        while a <= b:
            k = a * a + b * b

            if k < c:
                a += 1
            elif k > c:
                b -= 1
            else:
                return True

        return False



# class Solution:
#     def judgeSquareSum(self, c: int) -> bool:
#         for a in range(int(c**0.5) + 1):
#             x = c - a * a
#             if self.binarySearch(0, x, x):
#                 return True
#         return False

#     def binarySearch(self, b_start: int, b_end: int, x: int) -> bool:
#         while b_start <= b_end:
#             mid = b_start + (b_end - b_start) // 2
#             if mid * mid == x:
#                 return True
#             elif mid * mid < x:
#                 b_start = mid + 1
#             else:
#                 b_end = mid - 1
#         return False



# class Solution:
#     def judgeSquareSum(self, c: int) -> bool:
#         a = 0
#         while a * a <= c:
#             b = sqrt(c - a * a)
#             if b == int(b):
#                 return True
#             a += 1
#         return False